File: ruby/quicksort.rb
#!/usr/bin/env ruby

def quicksort(arr)
  if arr.empty? then arr
  else
    pivot = arr[0]
    lower, upper = arr[1..-1].partition { |x| x < pivot }
    [quicksort(lower), pivot, quicksort(upper)].flatten
  end
end

_ = gets
arr = gets.split(' ').map(&:to_i)
sorted = quicksort(arr)
puts sorted.join(' ')

File: python/quicksort.py
import sys

def quicksort(arr):
  if not arr:
    return []
  else:
    pivot = arr[0]
    lower = [x for x in arr[1:] if x < pivot]
    upper = [x for x in arr[1:] if x >= pivot]
    return quicksort(lower) + [pivot] + quicksort(upper)

lines = sys.stdin.readlines()
arr = [int(x) for x in lines[1].split(' ')]
sorted = quicksort(arr)

print ' '.join(str(x) for x in sorted)
